### ISSUE TITLE
Unable to get linear index >= 10000

### ISSUE THREAD

**Issue by jaksky** [deeplearning4j/nd4j#316]
After fresh checkout as of today ofr Cannova, ND4J, DL4J this issue still exist.

While I was trying the Deepautoencoder tutorial http://deeplearning4j.org/deepautoencoder.html I was getting following exception

Exception in thread "main" java.lang.IllegalArgumentException: Unable to get linear index >= 10000
at org.nd4j.linalg.api.ndarray.BaseNDArray.getDouble(BaseNDArray.java:3388)
at org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner.apply(DefaultOpExecutioner.java:531)
at org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner.exec(DefaultOpExecutioner.java:59)
at org.nd4j.linalg.api.ndarray.BaseNDArray.subi(BaseNDArray.java:2821)
at org.nd4j.linalg.api.ndarray.BaseNDArray.subi(BaseNDArray.java:2801)
at org.nd4j.linalg.api.ndarray.BaseNDArray.sub(BaseNDArray.java:2570)
at org.deeplearning4j.nn.layers.OutputLayer.backwardGradient(OutputLayer.java:96)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.doBackWard(MultiLayerNetwork.java:1086)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1255)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1283)

**Comment by agibsonccc** [comment]
@chrisvnicholson Can you reproduce this?

**Comment by jaksky** [comment]
Hi,

I did update to the latest code again and error is still there. Happens longer period of training (on my machine 15 min). As I wrote I was trying deep autoencoder example from the DL4J blog. I didn§t find that example in dl4j-examples so I tried following.

final int numRows = 28;
        final int numColumns = 28;
        int outputNum = 10;
        int seed = 123;
        int numSamples = MnistDataFetcher.NUM_EXAMPLES;
        int batchSize = 1000;
        int iterations = 5;
        int listenerFreq = iterations/5;

        log.info("Load data....");
        DataSetIterator iter = new MnistDataSetIterator(batchSize,numSamples,true);

        log.info("Build model....");
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .weightInit(WeightInit.DISTRIBUTION).dist(new NormalDistribution(0,0.01))
                .seed(seed)
                .constrainGradientToUnitNorm(true)
                .iterations(iterations)
                .updater(Updater.ADAGRAD)
                .momentum(0.5)
                .momentumAfter(Collections.singletonMap(3, 0.9))
                .optimizationAlgo(OptimizationAlgorithm.CONJUGATE_GRADIENT)
                .visibleUnit(RBM.VisibleUnit.BINARY)
                .hiddenUnit(RBM.HiddenUnit.BINARY)
                .list(10)
                .layer(0, new RBM.Builder().nIn(numRows*numColumns).nOut(1000).build())
                .layer(1, new RBM.Builder().nIn(1000).nOut(500).build())
                .layer(2, new RBM.Builder().nIn(500).nOut(250).build())
                .layer(3, new RBM.Builder().nIn(250).nOut(100).build())
                .layer(4, new RBM.Builder().nIn(100).nOut(30).build()) //encoding stops
                .layer(5, new RBM.Builder().nIn(30).nOut(100).build()) //decoding starts
                .layer(6, new RBM.Builder().nIn(100).nOut(250).build())
                .layer(7, new RBM.Builder().nIn(250).nOut(500).build())
                .layer(8, new RBM.Builder().nIn(500).nOut(1000).build())
                .layer(9, new OutputLayer.Builder(LossFunctions.LossFunction.RMSE_XENT).nIn(1000).nOut(numRows*numColumns).build())
                .pretrain(true).backprop(true)
                .build();

        MultiLayerNetwork model = new MultiLayerNetwork(conf);
        model.init();

        model.setListeners(Arrays.asList((IterationListener) new ScoreIterationListener(listenerFreq)));

        log.info("Train model....");
        while(iter.hasNext()) {
            DataSet next = iter.next();
            next.normalizeZeroMeanZeroUnitVariance();
            model.fit(next);
        }


        log.info("Evaluate model....");
        iter = new MnistDataSetIterator(100,100,true);
        DataSet dataSet = iter.next();
        Evaluation eval = new Evaluation();
        INDArray output = model.output(dataSet.getFeatureMatrix());
        eval.eval(dataSet.getLabels(), output);
        log.info(eval.stats());
    }

Hope this helps to troubleshoot this one.

Thanks for amazing project!

**Comment by agibsonccc** [comment]
This is because the "labels" on the dataset are the supervised learning labels for classification not reconstructing the input. See this example: https://github.com/deeplearning4j/dl4j-0.4-examples/commit/dbbd7f3062aaa1d99a2dd9bff88acc7e88b7dddd

**Comment by vallettea** [comment]
I have the same issue and I don't get the answer. Can you be a bit more specific ?

**Comment by agibsonccc** [comment]
Datasets are a pair of inputs and labels. You need to set the dataset as:
DataSet d = new DataSet(input,input);

The inputs are the labels here rather than a supervised classifier (which is the default mode)

**Comment by vallettea** [comment]
thanks a lot.
